File: iosApp/iosApp/Modules/ImprovementProposal/ImprovementProposalWireframeFactory.swift
// Created by web3d4v on 31/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit

protocol ImprovementProposalWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: ImprovementProposalWireframeContext
    ) -> ImprovementProposalWireframe
}

final class DefaultImprovementProposalWireframeFactory: ImprovementProposalWireframeFactory {
    
    func make(
        _ parent: UIViewController?,
        context: ImprovementProposalWireframeContext
    ) -> ImprovementProposalWireframe {
        DefaultImprovementProposalWireframe(parent, context: context)
    }
}

// MARK: - Assembler

final class ProposalWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> ImprovementProposalWireframeFactory in
            DefaultImprovementProposalWireframeFactory()
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/improvementproposal/ImprovementProposalWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.improvementproposal

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.improvementProposal.ImprovementProposalWireframe
import com.sonsofcrypto.web3walletcore.modules.improvementProposal.ImprovementProposalWireframeContext

interface ImprovementProposalWireframeFactory {
    fun make(
        parent: Fragment?,
        context: ImprovementProposalWireframeContext,
    ): ImprovementProposalWireframe
}

class DefaultImprovementProposalWireframeFactory(
): ImprovementProposalWireframeFactory {

    override fun make(
        parent: Fragment?,
        context: ImprovementProposalWireframeContext,
    ): ImprovementProposalWireframe {

        return DefaultImprovementProposalWireframe(
            parent?.let { WeakRef(it) },
            context
        )
    }
}

class ImprovementProposalWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register("ImprovementProposalWireframeFactory", AssemblerRegistryScope.INSTANCE) {
            DefaultImprovementProposalWireframeFactory()
        }
    }
}